// Copyright 2019 Zeppelin Bend Pty Ltd
// This file is part of cimbend.
// 
// cimbend is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// cimbend is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with cimbend.  If not, see <https://www.gnu.org/licenses/>.


syntax = "proto3";
option java_multiple_files = true;
option java_package = "com.zepben.protobuf.cim.iec61970.base.diagramlayout";
package zepben.cim.iec61970.base.diagramlayout;

import "zepben/cim/iec61970/base/diagramlayout/DiagramObjectPoint.proto";
import "zepben/cim/iec61970/base/diagramlayout/DiagramObjectStyle.proto";

/**
 * An object that defines one or more points in a given space. This object can be associated with anything
 * that specializes IdentifiedObject. For single line diagrams such objects typically include such items as
 * analog values, breakers, disconnectors, power transformers, and transmission lines.
 */
message DiagramObject {

    /**
     * Master resource identifier issued by a model authority. The mRID is unique within an exchange context. Global uniqueness
     * is easily achieved by using a UUID,  as specified in RFC 4122, for the mRID. The use of UUID is strongly recommended.
     * <p>
     * For CIMXML data files in RDF syntax conforming to IEC 61970-552 Edition 1, the mRID is mapped to rdf:ID or rdf:about attributes
     * that identify CIM object elements.
     */
    string mRID = 1;

    /**
     * The name is any free human readable and possibly non unique text naming the object.
     */
    string name = 2;

    /**
     * A diagram object is part of a diagram.
     */
    string diagramMRID = 3;

    /**
     * The domain object to which this diagram object is associated.
     */
    string identifiedObjectMRID = 4;

    /**
     * A diagram object has a style associated that provides a reference for the style used in the originating system.
     */
    iec61970.base.diagramlayout.DiagramObjectStyle diagramObjectStyle = 5;

    /**
     * Sets the angle of rotation of the diagram object.  Zero degrees is pointing to the top of the diagram.  Rotation is clockwise.
     */
    double rotation = 6;

    /**
     * The diagram objects that are associated with the domain object.
     * <p>
     * The diagram objects are returned in a map sorted by the diagram they live in.
     */
    repeated iec61970.base.diagramlayout.DiagramObjectPoint diagramObjectPoints = 16;

}
